File: 40-combination-sum-ii/40-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        
        candidates.sort()
        output = []
        def backtrack(path, index, target):
            if target == 0:
                output.append(path[:])
            for i in range(index, len(candidates)):
                if i > index and candidates[i] == candidates[i-1]:
                    continue
                if candidates[i] > target: break
                backtrack(path + [candidates[i]] , i+1,  target - candidates[i])

        
        backtrack([], 0, target)
        return output
            
        
                
                    

File: 40-combination-sum-ii/40-combination-sum-ii.cpp
class Solution {
public:
    
    void backtrack(vector<int>& grid, vector<int>& path, vector<vector<int>>& output, int target, int index, int s)
    {
        if(s == target)
        {
            output.push_back(path);
            return;
        }
        for(int i = index; i < grid.size(); i++)
        {
            if(i > index && grid[i] == grid[i-1])
                continue;
            if(s + grid[i] > target)
                continue;
            path.push_back(grid[i]);
            backtrack(grid, path, output, target, i + 1, s + grid[i]);
            path.pop_back();
        }
            
    }
    
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) 
    {
        sort(candidates.begin(), candidates.end());
        vector<int> path;
        vector <vector<int>> output;
        backtrack(candidates, path, output, target, 0, 0);
        return output;
    }
};